flips/20201310-storage-fees.md

```diff
@@ -0,0 +1,316 @@
+# Storage Fees
+
+| Status        | Proposed                                             |
+:-------------- |:---------------------------------------------------- |
+| **Author(s)** | Janez Podhostnik (janez.podhostnik@dapperlabs.com)   |
+| **Sponsor**   | Janez Podhostnik (janez.podhostnik@dapperlabs.com)   |
+| **Updated**   | 2020-10-13                                           |
+
+## Objective
+
+Limit the maximum amount of storage each account has, to prevent storing huge amount of data on chain.
+
+A minimum amount of storage will be provided to each account when it is created. More storage can then be purchased by using the **StorageFeesContract**. Storage can also be refunded using the same contract.
+
+## Motivation
+
+If execution state grows to large execution node will no longer be able to hold it in memory to calculate the next state. When this will happen the execution node requirements (and the nodes them selves) will need to be updated (and will cost more to run).
+
+Limiting the amount of storage an account can have will prevent the execution state growing because of two reasons:
+
+- attacks storing large amounts of data to the blockchain with the explicit intent to slow it down or crash it
+- developers not considering storage costs in their smart contracts (because they don't need to)
+
+## User Benefit
+
+This will have the long term benefit of reducing growth of the execution state size, and thus reducing growth of the execution node requirements. 
+
+Prescribing a reasonable minimum amount of storage will still allow almost everyone to deploy and use smart contracts normally. 
+
+## Design Proposal
+
+### Overview
+
+#### Account Storage Tracking (Execution Environment)
+
+- Every account should have a `storage_capacity` and `storage_used` of type uint64 and unit of byte
+- Minimum amount of `storage_capacity` for each account is 100KB (as defined by the **StorageFeesContract**)
+- `storage_used` is the sum of data stored on ledger (sum of register value byte size owned by this account) (register keys are not part of storage used)
```

Comment:
In Cadence, there's currently no limit on the length of path identifiers, so yes, this could be potentially abused.

Reply:
I agree they probably should be, but here is the comparison of the two choices as I see them:

Include keys in storage used cons:
- future changes to the how register keys are structured will require a migration for storage used by accounts
- user has less control over storage used

Don't include keys in storage used cons:
- possible attacks by setting very long path identifiers

If we are not planning on enforcing any other limits on path identifiers (and keys in general) including keys is the obvious choice.

Reply:
@turbolent @dete Today I was in a call about cadence extensibility and at some point it came up that it would be good to know the storage size of a certain instance of a resource. If the key of the resource is a part of its size it is virtually impossible to implement a method like this, because storing it under a different path would change its storage size.

Maybe we should consider a different solution for limiting resource path length...

Reply:
On the technical product meeting we decided that because of the reasons mentioned above it is better not to include keys into the amount of storage used.

The attack vector with having large paths will be solved elsewhere.